source/key.ts

```diff
@@ -0,0 +1,366 @@
+ */
+const generateSymKey = (type: D4LKeyTypes, algorithm: AesKeyGenParams | HmacKeyGenParams | Pbkdf2Params = AES_GCM): Promise<D4LKey> =>
+    // Parameters:
+    // 1. Symmetric Encryption algorithm name and its requirements
+    // 2. Boolean indicating extractable. which indicates whether or not the raw keying
+    //    material may be exported by the application
+    //    (http://www.w3.org/TR/WebCryptoAPI/#dfn-CryptoKey-slot-extractable)
+    // 3. Usage of the key. (http://www.w3.org/TR/WebCryptoAPI/#cryptokey-interface-types)
+    crypto.subtle.generateKey(
+        algorithm,
+        true,
+        [ENCRYPT, DECRYPT],
+    )
+        .then(key => exportKey(key, type));
+
+
+/**
+ * Creates key out of the given buffer (aka password)
+ *
+ * @param {Uint8Array} masterKey
+ * @param {Uint8Array} [salt] - The salt used for deriving, zeroed if not provided
+ * @param {AesDerivedKeyParams} algorithm - The algorithm used for deriving. AES GCM by default
+ * @returns {Promise<CryptoKey | D4LKey>} Resolves to a CryptoKey derived from the masterKey
+ */
+const deriveKey = (masterKey: Uint8Array, salt: Uint8Array = new Uint8Array(16), algorithm: AesDerivedKeyParams = AES_GCM): Promise<CryptoKey | D4LKey> => {
+    const ITERATIONS = 10000;
+
+    if (isEdge) {
+        return new Promise((resolve, reject) =>
+            // promisify
+            pbkdf2.pbkdf2(
+                convertArrayBufferViewToString(masterKey),
+                Buffer.from(salt),
+                ITERATIONS,
+                32,     // 32 bytes = 256 bits = key size of AES 256, the mode of operation is not relevant
+                'sha256',
+                (err, key) => {
+                    if (err) {
+                        reject(err);
+                        return;
+                    }
+
+                    resolve(key);
+                },
+            ),
+        )
+            .then(convertArrayBufferViewToBase64)
+            .then(importSymKeyFromBase64);
+    }
+
+    return crypto.subtle.importKey(
+        'raw',
+        masterKey,
+        { name: 'PBKDF2' },
+        false,
+        ['deriveKey', 'deriveBits'],
+    )
+        .then(key =>
+            crypto.subtle.deriveKey(
+                {
+                    name: 'PBKDF2',
+                    salt,
+                    iterations: ITERATIONS,
+                    hash: 'SHA-256',
+                },
+                key,
+                algorithm,
+                true,
+                [ENCRYPT, DECRYPT],
+            ),
+        )
+        .then(key => exportKey(key, D4LKeyTypes.PASSWORD_KEY));
+};
+
+/**
+ * Creates a random public-private key pair
+ * @param {AsymKeyTypes} type - type of the key pair (app/user)
+ * @returns {Promise<D4LAsymKeysPair>} Resolves to an object containing a public and a private key as d4lKey objects
+ */
+const generateAsymKeyPair = (type: AsymKeyTypes): Promise<D4LAsymKeysPair> =>
```

Comment:
I changed the type of the parameter here as it was depending on the enum-objects that I removed (it was also wrongly documented):
```
@param {String} type - type of the key pair 'A' for App, 'U' for User
[...]
            exportKey(keyPair.publicKey, type.PUBLIC_KEY),
            exportKey(keyPair.privateKey, type.PRIVATE_KEY),
```